### PR TITLE
[Discussion] make promise CE behave like the JS promise

### PR DESCRIPTION
**Pull request by 0x53A**
ref https://github.com/fable-compiler/fable-powerpack/issues/59

__EDIT:__ I got it to work:

grafik

---------

The object returned from ``Delay`` is unstable - querying it multiple times will execute it multiple times.

The solution is to wrap that object in ``Run``.

I chose ``Promise.Create`` as a wrapper, because that means it is a _true_ JS promise and behaves exactly the same.

### CODE REVIEW

**Comment by forki** [comment]
Tbh I always thought we are already having js semantics O_o

**Comment by baronfel** [comment]
I agree with making it mimic JS semantics as much as possible, especially since we also support async for cold computations

**Comment by vbfox** [comment]
The funny thing is that I'm always using the Promise wrapper in Ionide codebase because it behave exactly like JS.

> If no one subscribes to the object returned by promise { }, then nothing should be executed

It's not the JS behavior... `new Promise` both create a promise and start it, whenever you subscribe to it or not.

TBH I would prefer the current behavior to an half-half solution

**Comment by MangelMaxime** [comment]
I agree with @vbfox here

**Comment by 0x53A** [comment]
You are right.

This will immediatly log:

```js
var p1 = new Promise(
    function(resolve, reject) {       
      console.log("Inside Promise");
      
      resolve(1);
  }
);
```

--------------

I also agree with @vbfox and @MangelMaxime , I want the promise to behave the same as JS.
I just misunderstood JS ;D

File: tests/PromiseTests.fs [hunk 1]
```diff
@@ -322,3 +322,21 @@ describe "Promise tests" <| fun _ ->
         p.``then``(fun _ -> promiseExecutionCount |> equal 1)
         p.``then``(fun _ -> promiseExecutionCount |> equal 1)
         p.``then``(fun _ -> promiseExecutionCount |> equal 1)
```
Comment (vbfox) [inline tests/PromiseTests.fs]: Question about the test framework, does it handle correctly that it might execute asynchronously and so equal will be called potentially later ?

**Comment by alexandru** [comment]
Noticed this on my Twitter feed.

I’m not a Fable or F# user, but played with them, might do so again in the future, so take my opinion with a grain of salt.

I’m the author of [Monix](https://monix.io), of [Funfix](https://github.com/funfix/funfix) and the current leader of [Cats-Effect](https://typelevel.org/cats-effect/).

Mentioning this because I’ve implemented the “IO monad” multiple times already and the notion of “lazy promise” is very dear to my heart.

So here’s my two cents on this matter:

1. The semantics of the JavaScript Promise aren’t a virtue; in fact for users of saner implementations, like Scala’s `Future`, the `Promise` in JavaScript [is broken](https://alexn.org/blog/2017/10/11/javascript-promise-leaks-memory.html)
2. Strict evaluation for Promise-like data structures is anti-FP because it breaks referential transparency
3. Parallelism / concurrency is error prone because it couples evaluation order with side effects, so you have to pay attention to evaluation order, which in the context of asynchronous computations it’s not something that’s easy to do

By contrast “lazy promises”:

1. Enable [real FP](https://alexn.org/blog/2017/10/15/functional-programming.html) which we want because equational reasoning is nice to have when dealing with concurrency 
2. Decouples evaluation order from the side effects, which really means that the parallelism and the concurrency become really explicit and not something that can happen by accident 
3. Enable features that with strict Promises you cannot have

That piece of code where the log gets printed multiple times? Hell yeah, that’s how sane implementations should behace by default.

For example a lazy promise implementation can be retried in case of failure, you can define reusable functions with complex retry logic that works for any such value, with exponential back-off or whatever, which with a strict Promise you’d have to work with parameterless functions (aka thunks), which are error prone, due to the aforementioned evaluation.

You can also do lazy and thread-safe memoization, making it behave like a strict Promise, except that the value is computed only on first use and then reused for subsequent calls. 

There’s a lot of value in not implementing the semantics of JavaScript’s Promise and that’s not a bad thing. 

Going to be blunt about it, but after experiencing other implementations and ways of doing things, I think JavaScript’s Promise is the perfect example of everything that’s wrong with JavaScript.

**Comment by 0x53A** [comment]
I think the big topic is user expectations.

If the CE is called ``promise``, then my first expectation, without reading any docs, is that it behaves exactly like the JS promise, for better or worse.

------

There have already been lengthy discussions about C# Task vs F# Async, with similar content as your post.

And you are probably right that a lazy promise may be objectively better.

But then it needs a different name. Either call it ``lazyPromise``, or use ``async`` (which already exists and works in Fable).

Basically what I want to say is, that if your thing is called Promise, and there is already an existing Promise concept in the ecosystem, then it better behave the same (broken) way.

**Comment by alfonsogarciacaro** [comment]
Thanks a lot for your input @alexandru, it's very valuable! I do agree that lazy promises have many benefits and that's why they were implemented like that in the first place.

But also understand the concerns about user expectations and this seems to be a problem of weak semantics. Maybe we could deprecate `promise` (adding an Obsolete attribute so it doesn't break people's code) and replace it with two implementations: `lazy/coldPromise` that behaves as it does now and `cached/hotPromise` that matches JS behaviour?

**Comment by alfonsogarciacaro** [comment]
I meant deprecating to remove the ambiguity of `promise` and make people be more explicit about the intentions using either `hotPromise` or `coldPromise`. 

I'm not sure if it's a good choice to rename the current CE to `lazyPromise` and give a new behavior to `promise`. But if that's the preferred solution for the majority we can do it.

**Comment by MangelMaxime** [comment]
> I meant deprecating to remove the ambiguity of `promise` and make people be more explicit about the intentions using either `hotPromise` or `coldPromise`.
> 
> I'm not sure if it's a good choice to rename the current CE to `lazyPromise` and give a new behavior to `promise`. But if that's the preferred solution for the majority we can do it.

We should avoid this situation it will lead to potential subtile bugs.

In `Elmish.HMR` 3.x we did remove completly `Program.withHMR` so it give you a compiler error and people still have a hard time doing the conversation.

So imagine, if this is the same API name but the internal behavior changed... :)

**Comment by 0x53A** [comment]
I think that there are already many subtle bugs because people expect ``promise`` to behave like js (as evidenced by this thread), but it currently doesn't.

So I think my current preference would be to:

* mark ``promise`` as deprecated (possibly error level) and point them to this thread.
* create a new builder ``jsPromise`` (name open to debate)
* maybe create a new builder ``lazyPromise``.

But I don't think we should create a ``lazyPromise``. If you have a ``Promise`` object, you never know if it is lazy or eager. And you can't really interact with the JS ecosystem, because as soon as you do, your nice lazy promise is converted to eager.

So if we want lazy futures, we should either use ``async``, or create a new builder that wraps funfix.

**Comment by MangelMaxime** [comment]
Hum @0x53A made me think...

I think we should just fix current `promise` implementation. We kind of all think that it's having the same behavior as in JS and also Fable do favor interop in general.

IHMO if someone wants to implement "lazyPromise" he can create a dedicated library for it.

**Comment by alfonsogarciacaro** [comment]
Ok, given that most of the people expects `promise` to behave like JS promises even if this is not the "right" behaviour, let's merge this :+1: I'll publish a new major version of the library. Thanks @0x53A!

**Comment by 0x53A** [comment]
Thanks all!
